errors when extendedDebugInfo switch is enabled, even for valid configs

in v1.2.1, enabling the system flag

-Dsun.io.serialization.extendedDebugInfo=true

which is intended to make Java serialization verbose, raises the following error instead

~~~~
Caused by: com.typesafe.config.ConfigException$BugOrBroken: com.typesafe.config.impl.SerializedConfigValue should not exist outside of serialization
  at com.typesafe.config.impl.SerializedConfigValue.shouldNotBeUsed(SerializedConfigValue.java:471)
  at com.typesafe.config.impl.SerializedConfigValue.unwrapped(SerializedConfigValue.java:482)
  at com.typesafe.config.impl.AbstractConfigValue.render(AbstractConfigValue.java:325)
  at com.typesafe.config.impl.AbstractConfigValue.render(AbstractConfigValue.java:321)
  at com.typesafe.config.impl.AbstractConfigValue.toString(AbstractConfigValue.java:282)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1418)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1177)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1547)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1508)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1431)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1177)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1547)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1508)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1431)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1177)
  at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:347)
~~~~

It looks like this is because the flag enables printing out the object name and then writing a string representation, outside the context of the `Config` class...

This debug flag is useful in the case when you have a lot of nested objects, and are trying to figure out which one is causing serialization to fail. However, this bug throws up this error even for valid instance of Config, hence making the flag unusable.

--- a/config/src/main/java/com/typesafe/config/impl/AbstractConfigValue.java
+++ b/config/src/main/java/com/typesafe/config/impl/AbstractConfigValue.java
@@ -312,7 +312,7 @@ abstract class AbstractConfigValue implements ConfigValue, MergeableValue {
     }
 
     @Override
-    public final String toString() {
+    public String toString() {
         StringBuilder sb = new StringBuilder();
         render(sb, 0, true /* atRoot */, null /* atKey */, ConfigRenderOptions.concise());
         return getClass().getSimpleName() + "(" + sb.toString() + ")";

--- a/config/src/main/java/com/typesafe/config/impl/SerializedConfigValue.java
+++ b/config/src/main/java/com/typesafe/config/impl/SerializedConfigValue.java
@@ -497,4 +497,32 @@ class SerializedConfigValue extends AbstractConfigValue implements Externalizabl
     protected SerializedConfigValue newCopy(ConfigOrigin origin) {
         throw shouldNotBeUsed();
     }
+
+    @Override
+    public final String toString() {
+        return getClass().getSimpleName() + "(value=" + value + ",wasConfig=" + wasConfig + ")";
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        // there's no reason we will ever call this equals(), but
+        // the one in AbstractConfigValue would explode due to
+        // calling unwrapped() above, so we just give some
+        // safe-to-call implementation to avoid breaking the
+        // contract of java.lang.Object
+        if (other instanceof SerializedConfigValue) {
+            return canEqual(other)
+                && (this.wasConfig == ((SerializedConfigValue) other).wasConfig)
+                && (this.value.equals(((SerializedConfigValue) other).value));
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public int hashCode() {
+        int h = 41 * (41 + value.hashCode());
+        h = 41 * (h + (wasConfig ? 1 : 0));
+        return h;
+    }
 }

